File: TrueShuffle/MainActivity.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Android.App;
using Android.Content;
using Android.Content.PM;
using Android.Graphics;
using Android.OS;
using Android.Runtime;
using Android.Support.V7.App;
using Android.Util;
using Android.Views;
using Android.Widget;
using SpotifyAPI.Web;
using SpotifyAPI.Web.Models;
using Xamarin.Essentials;
using AlertDialog = Android.App.AlertDialog;

namespace TrueShuffle
{
    [Activity(Label = "@string/app_name", Theme = "@style/AppTheme")]
    public class MainActivity : AppCompatActivity
    {
        private const int CountPerPage = 8;
        private readonly EventHandler[] _buttonEvents = new EventHandler[CountPerPage];
        private readonly ValueListener<State> _state = new ValueListener<State> {Value = State.Waiting};
        private readonly object _stateChangeLock = new object();
        private SpotifyWebAPI _api;
        private int _currentIndex;
        private string _lastError;
        private List<SimplePlaylist> _playlists;
        private const int ArtistLimit = 10;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            Platform.Init(this, savedInstanceState);
            SetContentView(Resource.Layout.activity_main);

            string token = GetSharedPreferences("SPOTIFY", 0).GetString("token", "");
            if (token != "")
            {
                _api = new SpotifyWebAPI {AccessToken = token, TokenType = "Bearer"};
            }
            else
            {
                Intent i = new Intent(this, typeof(LoginActivity));
                i.SetFlags(ActivityFlags.NewTask | ActivityFlags.ClearTask);
                StartActivity(i);
            }

            Button nextButton = FindViewById<Button>(Resource.Id.next_button);
            Button previousButton = FindViewById<Button>(Resource.Id.previous_button);
            nextButton.Click += OnNavigationButtonClick;
            previousButton.Click += OnNavigationButtonClick;

            _state.Action = OnStateChange;
        }

        protected override void OnResume()
        {
            base.OnResume();

            LoadPlaylists(0);
        }

        private void LoadPlaylists(int index)
        {
            _currentIndex = index;

            // get playlists
            if (_playlists == null)
            {
                _playlists = new List<SimplePlaylist>();
                try
                {
                    string id = _api.GetPrivateProfile().Id;
                    Paging<SimplePlaylist> subPlaylists;
                    int i = 0;
                    do
                    {
                        subPlaylists = _api.GetUserPlaylists(id, CountPerPage, i++ * CountPerPage);
                        _playlists.AddRange(subPlaylists.Items);
                    } while (subPlaylists.HasNextPage());
                }
                catch (Exception)
                {
                    TextView statusTextView = FindViewById<TextView>(Resource.Id.status_text_view);
                    statusTextView.Text = "Failed to load playlists";
                    
                    return;
                }
            }

            // filter playlists
            List<SimplePlaylist> displayPlaylists = _playlists
                .Where(p =>
                    !(p.Name.EndsWith(" Shuffle") &&
                      _playlists.Any(p2 => p2.Name == p.Name.Substring(0, p.Name.Length - 8))) &&
                    !(p.Name.EndsWith(" Restrict") &&
                      _playlists.Any(p2 => p2.Name == p.Name.Substring(0, p.Name.Length - 9))))
                .ToList();

            // display playlists
            GridLayout parentLayout = FindViewById<GridLayout>(Resource.Id.playlist_display);
            for (int i = _currentIndex * CountPerPage; i < (_currentIndex + 1) * CountPerPage; i++)
                if (i < displayPlaylists.Count)
                {
                    TextView text = (TextView) parentLayout.GetChildAt(i % CountPerPage * 2);
                    Button button = (Button) parentLayout.GetChildAt(i % CountPerPage * 2 + 1);

                    text.Text = displayPlaylists[i].Name;
                    string id = displayPlaylists[i].Id;

                    if (_buttonEvents[i % CountPerPage] != null)
                        button.Click -= _buttonEvents[i % CountPerPage];

                    void Event(object sender, EventArgs e)
                    {
                        OnSelectButtonClick(id);
                    }

                    button.Click += Event;
                    _buttonEvents[i % CountPerPage] = Event;

                    text.Visibility = ViewStates.Visible;
                    button.Visibility = ViewStates.Visible;
                }
                else
                {
                    parentLayout.GetChildAt(i % CountPerPage * 2).Visibility = ViewStates.Gone;
                    parentLayout.GetChildAt(i % CountPerPage * 2 + 1).Visibility = ViewStates.Gone;
                }

            // show or hind nav buttons
            Button nextButton = FindViewById<Button>(Resource.Id.next_button);
            nextButton.Visibility = (_currentIndex + 1) * CountPerPage < _playlists.Count
                ? ViewStates.Visible
                : ViewStates.Invisible;

            Button previousButton = FindViewById<Button>(Resource.Id.previous_button);
            previousButton.Visibility = _currentIndex > 0 ? ViewStates.Visible : ViewStates.Invisible;
        }

        private void OnSelectButtonClick(string playlistId)
        {
            AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(this, Resource.Style.DialogTheme);
            LayoutInflater inflater = (LayoutInflater) GetSystemService(LayoutInflaterService);
            View popupView = inflater.Inflate(Resource.Layout.popup, null);

            dialogBuilder.SetView(popupView);
            AlertDialog dialog = dialogBuilder.Create();
            dialog.Show();

            popupView.FindViewById<Button>(Resource.Id.shuffle_button).Click += (sender, e) =>
            {
                dialog.Dismiss();
                OnShuffleButtonClick(playlistId, ShuffleMode.Shuffle, popupView);
            };
            popupView.FindViewById<Button>(Resource.Id.restrict_button).Click += (sender, e) =>
            {
                dialog.Dismiss();
                OnShuffleButtonClick(playlistId, ShuffleMode.Restrict, popupView);
            };
            popupView.FindViewById<EditText>(Resource.Id.restrict_value).Text =
                GetSharedPreferences("SPOTIFY", 0).GetString("RESTRICT_VALUE", "10");
        }

        private async void OnShuffleButtonClick(string playlistId, ShuffleMode shuffleMode, View popup)
        {
            if (_state.Value != State.Waiting && _state.Value != State.Failed) return;

            _state.Value = State.LoadingTracks;

            Task<Exception> task = Task<Exception>.Factory.StartNew(() =>
            {
                try
                {
                    // get playlist
                    FullPlaylist playlist =
                        _api.GetPlaylist(playlistId, fields: "id,name,tracks.items(track.uri),tracks.total");

                    if (playlist.Tracks == null)
                        return new Exception("Failed to get playlist");

                    // get tracks
                    IList<FullTrack> tracks = Enumerable.Range(0, (int) Math.Ceiling(playlist.Tracks.Total / 100F))
                        .Select(i =>
                            _api.GetPlaylistTracks(playlist.Id, fields: "items.track(uri,artists.name)",
                                offset: i * 100))
                        .SelectMany(group => group.Items.Select(playlistTrack => playlistTrack.Track))
                        .ToList();

                    if (tracks.Count != playlist.Tracks.Total)
                        return new Exception("Failed to get all tracks in playlist");

                    // randomize track order
                    if (shuffleMode == ShuffleMode.Shuffle)
                        tracks.Shuffle();
                    else if (shuffleMode == ShuffleMode.Restrict)
                    {
                        // get value
                        EditText value = popup.FindViewById<EditText>(Resource.Id.restrict_value);
                        bool res = int.TryParse(value.Text, out int artistLimit);

                        // check value is valid
                        if(!res || artistLimit < 1) throw new Exception("Restrict value is not a positive integer");
                        
                        // save value
                        ISharedPreferencesEditor editor = GetSharedPreferences("SPOTIFY", 0).Edit();
                        editor.PutString("RESTRICT_VALUE", artistLimit.ToString());
                        editor.Commit();
                        
                        tracks = tracks
                            .GroupBy(track => track.Artists[0].Name)
                            .SelectMany(artist => artist.ToList().Shuffle().Take(artistLimit))
                            .ToList()
                            .Shuffle();
                    }

                    // add tracks
                    string id = _api.GetPrivateProfile().Id;

                    // delete old playlists
                    foreach (SimplePlaylist p in _playlists.Where(p => p.Name == playlist.Name + " " + shuffleMode)
                        .ToList())
                    {
                        ErrorResponse error = _api.UnfollowPlaylist(id, p.Id);
                        if (error.HasError())
                            return new Exception("Failed to delete old playlist");
                        _playlists.Remove(p);
                    }

                    // create new playlist
                    _state.Value = State.AddingTracks;
                    FullPlaylist newPlaylist = _api.CreatePlaylist(id, playlist.Name + " " + shuffleMode, false);
                    if (newPlaylist.Id == null)
                        return new Exception("Failed to create new playlist");


                    _playlists.Add(new SimplePlaylist {Id = newPlaylist.Id, Name = newPlaylist.Name});

                    // add tracks
                    List<ErrorResponse> addResult = Enumerable.Range(0, (int) Math.Ceiling(tracks.Count / 100F)).Select(
                            i =>
                                _api.AddPlaylistTracks(newPlaylist.Id,
                                    tracks.Skip(i * 100).Take(100).Select(track => track.Uri).ToList()))
                        .ToList();

                    if (addResult.Any(error => error.HasError()))
                        return new Exception("Failed to add all tracks to playlist");

                    return null;
                }
                catch (Exception)
                {
                    return new Exception("Unknown error has occured");
                }
            });

            await task.ContinueWith(errorTask =>
            {
                if (errorTask.Result == null)
                {
                    Log.Debug("ShuffleResult", "Shuffle succeeded");

                    _state.Value = State.Waiting;
                }
                else
                {
                    Log.Debug("ShuffleResult", $"Shuffle failed: {errorTask.Result}");
                    _lastError = errorTask.Result.Message;

                    _state.Value = State.Failed;
                }
            });
        }

        private void OnStateChange(object sender, ValueListenerEventArgs<State> e)
        {
            if (_state.Value == e.OldValue) return;

            RunOnUiThread(OnStateChangeUi);
        }

        private void OnStateChangeUi()
        {
            lock (_stateChangeLock)
            {
                TextView statusTextView = FindViewById<TextView>(Resource.Id.status_text_view);
                switch (_state.Value)
                {
                    case State.LoadingTracks:
                        statusTextView.Text = "Loading Tracks...";
                        break;
                    case State.RemovingTracks:
                        statusTextView.Text = "Removing Old Tracks...";
                        break;
                    case State.AddingTracks:
                        statusTextView.Text = "Adding Shuffled Tracks...";
                        break;
                    case State.Waiting:
                        statusTextView.Text = "";
                        break;
                    case State.Failed:
                        statusTextView.Text = _lastError;
                        break;
                    default:
                        throw new ArgumentOutOfRangeException();
                }


                GridLayout parentLayout = FindViewById<GridLayout>(Resource.Id.playlist_display);
                for (int i = 1; i < parentLayout.ChildCount; i += 2)
                {
                    Button button = (Button) parentLayout.GetChildAt(i);
                    button.Activated = _state.Value == State.Waiting || _state.Value == State.Failed;
                    Color color = new Color(GetColor(_state.Value == State.Waiting || _state.Value == State.Failed
                        ? Resource.Color.spotify_green
                        : Resource.Color.spotify_grey));
                    button.SetBackgroundColor(color);
                }
            }
        }

        private void OnNavigationButtonClick(object sender, EventArgs e)
        {
            Button senderButton = (Button) sender;
            int offset = senderButton.Id == Resource.Id.next_button ? 1 : -1;
            LoadPlaylists(_currentIndex + offset);
        }

        public override void OnRequestPermissionsResult(int requestCode, string[] permissions,
            [GeneratedEnum] Permission[] grantResults)
        {
            Platform.OnRequestPermissionsResult(requestCode, permissions, grantResults);

            base.OnRequestPermissionsResult(requestCode, permissions, grantResults);
        }

        private enum State
        {
            LoadingTracks,
            RemovingTracks,
            AddingTracks,
            Waiting,
            Failed
        }

        private enum ShuffleMode
        {
            Shuffle,
            Restrict
        }
    }
}

File: app/src/main/java/com/csalisbury/trueshuffle/MainActivity.java
package com.csalisbury.trueshuffle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.GridLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.spotify.sdk.android.auth.AuthorizationClient;
import com.spotify.sdk.android.auth.AuthorizationRequest;
import com.spotify.sdk.android.auth.AuthorizationResponse;

import java.security.InvalidParameterException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.Semaphore;
import java.util.stream.Collectors;

import kaaes.spotify.webapi.android.SpotifyApi;
import kaaes.spotify.webapi.android.SpotifyService;
import kaaes.spotify.webapi.android.models.Pager;
import kaaes.spotify.webapi.android.models.PlaylistBase;
import kaaes.spotify.webapi.android.models.PlaylistSimple;
import kaaes.spotify.webapi.android.models.PlaylistTrack;
import kaaes.spotify.webapi.android.models.TrackToRemove;
import kaaes.spotify.webapi.android.models.TracksToRemove;
import kaaes.spotify.webapi.android.models.UserPrivate;
import retrofit.RetrofitError;

public class MainActivity extends AppCompatActivity {
    private static final String CLIENT_ID = "c4a1a641314d4d99b9df5304449a6c8d";
    private static final String REDIRECT_URI = "com.csalisbury.trueshuffle://callback";
    private static final int REQUEST_CODE = 31415;
    private static final String[] SCOPES = new String[]{"playlist-read-private playlist-modify-public playlist-modify-private"};
    private static final int PLAYLISTS_PER_PAGE = 10;
    private List<PlaylistBase> playlists;
    private List<PlaylistBase> playlistChildren;
    private SpotifyApi api;
    private SpotifyService spotify;
    private UserPrivate me;
    private int playlistIndex;
    private SwipeRefreshLayout swipe;
    private final Semaphore semaphore = new Semaphore(1);
    private ValueWatcher<State> currentState;
    private final ShuffleType[] SHUFFLE_TYPES = new ShuffleType[]{
            new ShuffleType("Shuffle", "Shuffled", R.id.shuffle_button, this::shuffleAll),
            new ShuffleType("Limit", "Limited", R.id.restrict_button, R.id.restrict_value, this::shuffleLimited)
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    @Override
    protected void onStart() {
        super.onStart();

        currentState = new ValueWatcher<>(new State(State.Type.OK), s -> runOnUiThread(() -> onStateChange(s)));

        login();

        findViewById(R.id.next_button).setOnClickListener(v -> populateUI(++playlistIndex * PLAYLISTS_PER_PAGE));
        findViewById(R.id.previous_button).setOnClickListener(v -> populateUI(--playlistIndex * PLAYLISTS_PER_PAGE));
        swipe = findViewById(R.id.refresh_view);
        swipe.setOnRefreshListener(this::onRefreshPage);
        findViewById(R.id.status_text_view).setVisibility(View.GONE);
    }

    private void onRefreshPage() {
        try {
            semaphore.acquire();
        } catch (Exception e) {
            return;
        }

        if (currentState.getValue().isReadyState()) {
            currentState.setValue(new State(State.Type.PROCESSING, "Refreshing playlists"));
            semaphore.release();
            loadPlaylists();
        } else if (currentState.getValue().type == State.Type.LOGIN_FAILED) {
            currentState.setValue(new State(State.Type.PROCESSING, "Logging in"));
            semaphore.release();
            login();
        } else {
            semaphore.release();
            currentState.setValue(new State(currentState.getValue().type, "Cannot refresh at the while something else is happening"));
            swipe.setRefreshing(false);
        }
    }

    private void login() {
        AuthorizationRequest.Builder builder = new AuthorizationRequest.Builder(CLIENT_ID, AuthorizationResponse.Type.TOKEN, REDIRECT_URI);

        builder.setScopes(SCOPES);
        AuthorizationRequest request = builder.build();

        AuthorizationClient.openLoginActivity(this, REQUEST_CODE, request);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == REQUEST_CODE) {
            AuthorizationResponse response = AuthorizationClient.getResponse(resultCode, data);

            switch (response.getType()) {
                case TOKEN:
                    Log.d("SpotifyAuth", "Auth token: " + response.getAccessToken());
                    currentState.setValue(new State(State.Type.OK));
                    setupApi(response.getAccessToken());
                    loadPlaylists();
                    break;
                case ERROR:
                    Log.d("SpotifyAuth", "Auth error: " + response.getError());
                    currentState.setValue(new State(State.Type.LOGIN_FAILED, "Failed to login"));
                    break;
                default:
                    Log.d("SpotifyAuth", "Auth aborted: " + response.getError());
                    currentState.setValue(new State(State.Type.LOGIN_FAILED, "Failed to login"));
            }
        }
    }

    private void setupApi(String accessToken) {
        api = new SpotifyApi();

        api.setAccessToken(accessToken);

        spotify = api.getService();
    }

    private void loadPlaylists() {
        new Thread(() -> {
            LinkedList<PlaylistBase> playlists = new LinkedList<>();
            Pager<PlaylistSimple> playlistResult;
            int i = 0;
            try {
                do {
                    playlistResult = spotify.getMyPlaylists(new Options(String.format("offset=%s limit=%s", i, 20)));
                    playlists.addAll(playlistResult.items);
                    i += 20;
                } while (playlistResult.next != null);
                currentState.setValue(new State(State.Type.OK));
            } catch (RetrofitError e) {
                Log.e("SpotifyAPI", "Error occurred fetching playlists: " + e);
                handleAPIError(e, "Failed to load playlists");
            }

            this.playlists = playlists.stream().filter(p -> {
                for (ShuffleType s : SHUFFLE_TYPES) {
                    if (p.name.endsWith(" " + s.getSuffix()))
                        return false;
                }
                return true;
            }).collect(Collectors.toList());

            this.playlistChildren = playlists.stream().filter(p -> {
                for (ShuffleType s : SHUFFLE_TYPES) {
                    if (p.name.endsWith(" " + s.getSuffix()))
                        return true;
                }
                return false;
            }).collect(Collectors.toList());

            playlistIndex = 0;
            runOnUiThread(() -> populateUI(0));
        }).start();
    }

    private void populateUI(int offset) {
        GridLayout parentLayout = findViewById(R.id.playlist_display);
        for (int i = offset; i < offset + PLAYLISTS_PER_PAGE; i++)
            if (i < playlists.size()) {
                TextView text = (TextView) parentLayout.getChildAt(i % PLAYLISTS_PER_PAGE * 2);
                Button button = (Button) parentLayout.getChildAt(i % PLAYLISTS_PER_PAGE * 2 + 1);

                text.setText(playlists.get(i).name);
                PlaylistBase playlist = playlists.get(i);

                button.setOnClickListener(v -> onDisplayShuffleDialog(playlist));

                text.setVisibility(View.VISIBLE);
                button.setVisibility(View.VISIBLE);

            } else {
                parentLayout.getChildAt(i % PLAYLISTS_PER_PAGE * 2).setVisibility(View.GONE);
                parentLayout.getChildAt(i % PLAYLISTS_PER_PAGE * 2 + 1).setVisibility(View.GONE);
            }

        Button nextButton = findViewById(R.id.next_button);
        nextButton.setVisibility(offset + PLAYLISTS_PER_PAGE < playlists.size()
                ? View.VISIBLE
                : View.INVISIBLE);

        Button previousButton = findViewById(R.id.previous_button);
        previousButton.setVisibility(offset > 0 ? View.VISIBLE : View.INVISIBLE);
        swipe.setRefreshing(false);
    }

    private void onDisplayShuffleDialog(PlaylistBase playlist) {
        try {
            semaphore.acquire();
        } catch (Exception e) {
            return;
        }
        if (!currentState.getValue().isReadyState())
            return;

        currentState.setValue(new State(State.Type.PROCESSING));
        semaphore.release();

        LayoutInflater inflater = (LayoutInflater) getSystemService(LAYOUT_INFLATER_SERVICE);
        View popupView = inflater.inflate(R.layout.shuffle_popup, null);

        new Thread(() -> runOnUiThread(() -> {
            for (ShuffleType s : SHUFFLE_TYPES) {
                if (!s.takesInput()) continue;
                String value = getSharedPreferences("SPOTIFY", MODE_PRIVATE).getString(s.getName() + "_input", "");
                if (value.equals("")) continue;
                ((EditText) popupView.findViewById(s.getInputId())).setText(value);
            }
        })).start();

        int width = LinearLayout.LayoutParams.MATCH_PARENT;
        int height = LinearLayout.LayoutParams.WRAP_CONTENT;
        PopupWindowMod pw = new PopupWindowMod(popupView, width, height, true);
        pw.setBackgroundDrawable(new ColorDrawable(ContextCompat.getColor(this, R.color.spotify_grey)));
        pw.showAtLocation(swipe, Gravity.CENTER, 0, 0);
        pw.setOnDismissListener(() -> {
            if (pw.isCancelled()) {
                currentState.setValue(new State(State.Type.OK));
            }
        });


        for (ShuffleType s : SHUFFLE_TYPES) {
            Button button = popupView.findViewById(s.getButtonId());
            button.setText(s.getName());
            button.setOnClickListener(v -> {
                pw.dismiss(false);

                String input = null;
                if (s.takesInput()) {
                    EditText et = popupView.findViewById(s.getInputId());
                    input = et.getText().toString();
                    SharedPreferences.Editor edit = getSharedPreferences("SPOTIFY", MODE_PRIVATE).edit();
                    edit.putString(s.getName() + "_input", input);
                    edit.apply();
                }
                shuffle(playlist, s, input);
            });
        }

        ((EditText) popupView.findViewById(R.id.restrict_value)).setText(
                getSharedPreferences("SPOTIFY", 0).getString("RESTRICT_VALUE", "10"));
    }

    private void shuffle(PlaylistBase playlist, ShuffleType shuffleType, String input) {
        new Thread(() -> {
            currentState.setValue(new State(State.Type.PROCESSING, "Loading songs"));

            List<PlaylistTrack> tracks;
            try {
                tracks = getAllTracks(playlist, true);
            } catch (RetrofitError e) {
                Log.e("SpotifyAPI", "Error occurred fetching tracks: " + e);
                handleAPIError(e, "Failed to load songs");
                return;
            }

            tracks = shuffleType.shuffle(tracks, input);

            PlaylistBase playlistChild = null;
            for (PlaylistBase p : playlistChildren) {
                if (p.name.equals(playlist.name + " " + shuffleType.getSuffix())) {
                    playlistChild = p;

                    currentState.setValue(new State(State.Type.PROCESSING, "Deleting old songs"));
                    try {
                        List<PlaylistTrack> existingTracks = getAllTracks(playlistChild, false);
                        deleteAllTracks(playlistChild, existingTracks);
                    } catch (RetrofitError e) {
                        Log.e("SpotifyAPI", "Error occurred deleting tracks: " + e);
                        handleAPIError(e, "Failed to delete old songs");
                        return;
                    }

                    break;
                }
            }

            if (playlistChild == null) {
                currentState.setValue(new State(State.Type.PROCESSING, "Creating new playlist"));
                try {
                    playlistChild = spotify.createPlaylist(playlist.owner.id, new Options("").add("name", String.format("%s %s", playlist.name, shuffleType.getSuffix())).add("public", false));
                } catch (RetrofitError e) {
                    Log.e("SpotifyAPI", "Error occurred creating new playlist: " + e);
                    handleAPIError(e, "Failed to create new playlist");
                    return;
                }
                playlistChildren.add(playlistChild);
            }

            currentState.setValue(new State(State.Type.PROCESSING, "Adding songs"));
            try {
                addTracks(playlistChild, tracks);
            } catch (RetrofitError e) {
                Log.e("SpotifyAPI", "Error occurred adding tracks: " + e);
                handleAPIError(e, "Failed to add songs");
                return;
            }

            currentState.setValue(new State(State.Type.OK));
        }).start();
    }

    private List<PlaylistTrack> getAllTracks(PlaylistBase playlist, boolean requireArtists) {
        int trackCount = spotify.getPlaylist(playlist.owner.id, playlist.id, new Options("fields=tracks.total")).tracks.total;

        String fields = requireArtists
                ? "items.track(uri,artists.name)"
                : "items.track(uri)";

        List<PlaylistTrack> tracks = new ArrayList<>();
        for (int i = 0; i < (int) Math.ceil(trackCount) / 100f; i++) {
            Pager<PlaylistTrack> playlistTracks = spotify.getPlaylistTracks(playlist.owner.id, playlist.id, new Options(String.format("limit=100 offset=%s fields=%s", i * 100, fields)));
            tracks.addAll(playlistTracks.items);
        }
        return tracks;
    }

    private void deleteAllTracks(PlaylistBase playlist, List<PlaylistTrack> tracks) {
        for (int i = 0; i < (int) Math.ceil(tracks.size()) / 100f; i++) {
            TracksToRemove tracksToRemove = new TracksToRemove();
            tracksToRemove.tracks = tracks.stream()
                    .skip(i * 100L)
                    .limit(100)
                    .map(t -> {
                        TrackToRemove ttr = new TrackToRemove();
                        ttr.uri = t.track.uri;
                        return ttr;
                    })
                    .collect(Collectors.toList());

            spotify.removeTracksFromPlaylist(playlist.owner.id, playlist.id, tracksToRemove);
        }
    }

    private void addTracks(PlaylistBase playlist, List<PlaylistTrack> tracks) {
        for (int i = 0; i < (int) Math.ceil(tracks.size()) / 100f; i++) {
            TracksToRemove tracksToRemove = new TracksToRemove();
            tracksToRemove.tracks = tracks.stream()
                    .skip(i * 100L)
                    .limit(100)
                    .map(t -> {
                        TrackToRemove ttr = new TrackToRemove();
                        ttr.uri = t.track.uri;
                        return ttr;
                    })
                    .collect(Collectors.toList());

            spotify.addTracksToPlaylist(playlist.owner.id, playlist.id,
                    new Options(""),
                    new Options("position=" + (i * 100))
                            .add("uris", tracks.stream().skip(i * 100L).limit(100).map(t -> t.track.uri).toArray())
            );
        }
    }

    private void onStateChange(State s) {
        TextView v = findViewById(R.id.status_text_view);
        if (!s.message.equals("")) {
            v.setText(s.message);
            v.setVisibility(View.VISIBLE);
        } else {
            v.setVisibility(View.GONE);
        }

        GridLayout parentLayout = findViewById(R.id.playlist_display);
        for (int i = 1; i < parentLayout.getChildCount(); i += 2) {
            Button button = (Button) parentLayout.getChildAt(i);
            boolean active = currentState.getValue().isReadyState();
            button.setActivated(active);
            int color = active
                    ? R.color.spotify_green
                    : R.color.spotify_grey;
            button.setBackgroundColor(ContextCompat.getColor(this, color));
        }
    }

    private void handleAPIError(RetrofitError e, String defaultMessage) {
        int resCode = e.getResponse().getStatus();
        switch (resCode) {
            case 401:
                currentState.setValue(new State(State.Type.LOGIN_FAILED, "Session expires, please refresh page to log back in"));
                break;
            case 403:
                currentState.setValue(new State(State.Type.ERROR, "Rate limit exceeded"));
                break;
            default:
                currentState.setValue(new State(State.Type.ERROR, defaultMessage));
        }
    }

    private List<PlaylistTrack> shuffleAll(List<PlaylistTrack> tracks, String input) {
        List<PlaylistTrack> result = new ArrayList<>(tracks);
        randomizeList(result);
        return result;
    }

    private List<PlaylistTrack> shuffleLimited(List<PlaylistTrack> tracks, String input) {
        int maxPerArtist;
        try {
            maxPerArtist = Integer.parseInt(input);
            if (maxPerArtist <= 0) throw new Exception();
        } catch (Exception e) {
            currentState.setValue(new State(State.Type.ERROR, "Input must be a positive"));
            throw new InvalidParameterException(String.format("Input value \"%s\" is not a positive integer.", input));
        }

        List<PlaylistTrack> result = new ArrayList<>();
        Map<String, List<PlaylistTrack>> tracksByArtist = tracks.stream().collect(Collectors.groupingBy(t -> t.track.artists.get(0).name));
        tracksByArtist.forEach((artist, ts) -> {
            randomizeList(ts);
            result.addAll(ts.stream().limit(maxPerArtist).collect(Collectors.toList()));
        });

        randomizeList(result);
        return result;
    }

    private static <T> void randomizeList(List<T> list) {
        Random rnd = new Random();
        for (int i = 0; i < list.size() - 2; i++) {
            int j = rnd.nextInt(list.size() - i) + i;
            T temp = list.get(i);
            list.set(i, list.get(j));
            list.set(j, temp);
        }
    }
}